### PR TITLE
CDAP-6008,CDAP-4647 Mark Workflow as failed if action executed by the Workflow fails.

### PR DESCRIPTION
**Pull request by sagarkapare**
JIRA: https://issues.cask.co/browse/CDAP-6008
JIRA: https://issues.cask.co/browse/CDAP-4647

### CODE REVIEW

**Comment by sagarkapare** [comment]
Release build running here - http://builds.cask.co/browse/CDAP-RUT51-1. Testing on the cluster.

**Comment by sagarkapare** [comment]
Verified on the cluster. If action executed by Workflow fails, Workflow status is marked as failed correctly. @chtyim Please take a look.

New build is running here - http://builds.cask.co/browse/CDAP-RUT51-11. There are some failures I am seeing on the bamboo boxes - 
1. AdminAppTestRun#testAdminWorkflow 
2. TestFrameworkTestRun#testAppWithPlugin
3. TestFrameworkTestRun#testWorkflowStatus

These tests are passing locally. Will need to investigate further.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowProgramController.java [hunk 1]
```diff
@@ -81,10 +81,13 @@ public void running() {
 
       @Override
       public void terminated(Service.State from) {
-        LOG.info("Workflow service terminated from {}. Un-registering service {}.", from, serviceName);
-        cancelAnnounce.cancel();
+        LOG.info("Workflow service terminated from {}. Un-registering service {}. getState: {}", from, serviceName,
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowProgramController.java:83]: s/getState/Workflow state/

**Comment by sreevatsanraman** [comment]
LGTM. Merge once the build passes.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowProgramRunner.java [hunk 2]
```diff
@@ -101,25 +115,100 @@ public ProgramController run(Program program, ProgramOptions options) {
...
+            startTimeInSeconds = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
+          }
+          runtimeStore.setStart(program.getId(), runId.getId(), startTimeInSeconds, twillRunId,
+                                options.getUserArguments().asMap(), options.getArguments().asMap());
+          if (state == ProgramController.State.COMPLETED) {
+            completed();
+          }
+          if (state == ProgramController.State.ERROR) {
+            error(controller.getFailureCause());
+          }
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowProgramRunner.java:170]: Isn't that you should also handle the suspended and the alive state (after resuming is alive and you get resuming before it actually resumes).

**Comment by sagarkapare** [comment]
Closing this PR.

**Pull request closed without merging**